### PR TITLE
feat: check INSTALL.md (TRG 1.02)

### LINKED ISSUE #10: Add check for existence of INSTALL.md
# What

In [TRG-1.02](https://eclipse-tractusx.github.io/docs/release/trg-1/trg-1-2) , requirements for an `INSTALL.md` file is described. 
So a QualityGuideline testing for the existence of this file should be added and run in the `checkLocal` command

# Discussion

This TRG is an optional one. The `INSTALL.md` file _should_ be present, if installation instructions in the `README.md` are not sufficient, or are moved to a dedicated file, because of a better overview. This criteria can hardly be tested automatically.

Currently, the checks are implemented with either fail or pass. There is no concept yet for 'optional' requirements.
In my opinion, we should discuss: 

- What should happen, if an optional requirement is not met. 
  - Pass, but still log warnings? 
  - Fail, with a hint, that with additional review it can be ignored?

### PR DESCRIPTION
**Pull request by carslen**
Implemented new method `IsOptional` in interface `QualityGuideline` to cover optional checks (aka `INSTALL.md` or `AUTHOR.md`(→ TRG 2.03)).

Added new feature to test if optional file 'INSTALL.md' exists. If the file does not exist, this test fails (for now) but not the test run itself (does not exit 1):

```shell
$ go run main.go checkLocal
Running local checks of eclipse-tractusx release guidelines
Testing Quality Guideline: TRG 1.01 - README.md
Testing Quality Guideline: TRG 1.02 - INSTALL.md
Failed! Test failed, but marked as optional. More infos: Did not find optional INSTALL.md file in current directory.
https://eclipse-tractusx.github.io/docs/release/trg-1/trg-1-2
Testing Quality Guideline: TRG 1.03 - CHANGELOG.md
```

Closes #10

### CODE REVIEW

**Comment by Siegfriedk** [comment]
Just a general sidenote: if we have more checks, should we try to prefix them or organize them more clearly?

**Comment by hzierer** [comment]
IMHO, I would rephrase the Message since "Failed! Test failed, but marked as optional." sounds quite harsh.
Maybe something like "Warning! Optional Test failed, but is considered best practice..."

**Comment by carslen** [comment]
> IMHO, I would rephrase the Message since "Failed! Test failed, but marked as optional." sounds quite harsh. Maybe something like "Warning! Optional Test failed, but is considered best practice..."

Technically the test fails, therefore wording is correct. But I get your point. Currently there is nothing between Passed and Failed implemented.

Would that be a better output:

```shell
$ go run main.go checkLocal
Running local checks of eclipse-tractusx release guidelines
Testing Quality Guideline: TRG 1.01 - README.md
Testing Quality Guideline: TRG 1.02 - INSTALL.md
Warning! Test failed, but test is marked as optional.
More infos:
        Optional INSTALL.md file not found in current directory.
        https://eclipse-tractusx.github.io/docs/release/trg-1/trg-1-2
Testing Quality Guideline: TRG 1.03 - CHANGELOG.md
```

**Comment by hzierer** [comment]
Maybe aligning it a bit more with the other messages:
Warning! Testing (Optional) Quality Guideline: TRG 1.02 - INSTALL.md failed, more info: ...

**Comment by carslen** [comment]
> Maybe aligning it a bit more with the other messages: Warning! Testing (Optional) Quality Guideline: TRG 1.02 - INSTALL.md failed, more info: ...

I'd leave it for not this way and redesign message handling with another PR.

**Comment by carslen** [comment]
New check already applied to this PR: https://github.com/eclipse-tractusx/tractusx-quality-checks/actions/runs/4628495387/jobs/8187637882

image

**Comment by Siegfriedk** [comment]
@carslen & @SebastianBezold can we make it more github action native? Like showing it in the checks instead of indirectly in the github action?

Also potentially you can indent it to make it easy seeable that the warning is part of TRG 1.02

**Comment by SebastianBezold** [comment]
> @carslen & @SebastianBezold can we make it more github action native? Like showing it in the checks instead of indirectly in the github action?
> 
> Also potentially you can indent it to make it easy seeable that the warning is part of TRG 1.02

In general, yes we can do that. In this case I would suggest to create a separate command, that "knows" or "assumes" it is running on PR basis and therefore has the GitHub features available. 
I would create it as separate commands, so we do not need to add too many conditions, checking in which context the checks are run.
The checks themselves run under the assumptions, that the files to check are available locally, relative to where the command is run from. So if we build something that matches this assumption and makes the results accessible via GitHub check API, then this can definitely be done

File: internal/install_exists.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+// InstallExists defines a struct for
+type InstallExists struct {
+}
+
+// NewInstallExists func return a new check based on QualityGuideline interface.
+func NewInstallExists() QualityGuideline {
+	return &InstallExists{}
+}
+
+// Name implements ExternalDescription from interface
```
Comment (SebastianBezold) [inline internal/install_exists.go:33]: I like, that you add docs for the functions/methods, but in this particular case, I would rather move it to the interface itself, since the intention is defined by the interface. The implementations then do not need to add it again and again, since `Name()` `Description` and `ExternalDescription()` do not really have specific implementation details that have to be described on top of the interface intend

**Review by hzierer — COMMENTED** [review]
Tested it successfully against 2 repos locally (1 positive, 1 negative)

File: internal/install_exists.go (hunk 1 above)
Comment (carslen) [reply internal/install_exists.go:33]: Jap, you're right. To be honest, I was pretty unsure about where to add docs. I'm going to change this.

Comment (carslen) [reply internal/install_exists.go]: Done

**Review by SebastianBezold — APPROVED** [review]
